pkg/codegen/java/gen_program.go

```diff
@@ -229,33 +229,334 @@ func containConfigVariables(nodes []pcl.Node) bool {
+					if innerObjectType.IsInputShape() {
+						objectTypeName = objectTypeName + "Args"
+						imports = append(imports, pulumiInputImport(pkg, module, objectTypeName))
+					} else {
+						imports = append(imports, pulumiImport(pkg, module, objectTypeName))
+					}
+
+					// recurse into nested object
+					for _, importDef := range collectObjectImports(innerObject, innerObjectType) {
+						imports = append(imports, importDef)
+					}
+				}
+			}
+		}
+	}
+
+	return imports
+}
+
+func collectResourceImports(resource *pcl.Resource) []string {
+	imports := make([]string, 0)
+	pkg, module, name, _ := resource.DecomposeToken()
+	resourceImport := pulumiImport(pkg, module, name)
+	imports = append(imports, resourceImport)
+	if len(resource.Inputs) > 0 || hasCustomResourceOptions(resource) {
+		// import args type name
+		argsTypeName := resourceArgsTypeName(resource)
+		resourceArgsImport := pulumiImport(pkg, module, argsTypeName)
+		imports = append(imports, resourceArgsImport)
+		resourceProperties := typedResourceProperties(resource)
+		for _, inputProperty := range resource.Inputs {
+			inputType := resourceProperties[inputProperty.Name]
+			switch inputType.(type) {
+			case *schema.ObjectType:
+				objectType := inputType.(*schema.ObjectType)
+				switch inputProperty.Value.(type) {
+				case *model.ObjectConsExpression:
+					object := inputProperty.Value.(*model.ObjectConsExpression)
+					fullyQualifiedTypeName := objectType.Token
+					nameParts := strings.Split(fullyQualifiedTypeName, ":")
+					objectTypeName := names.Title(nameParts[len(nameParts)-1])
+					if objectType.IsInputShape() {
+						objectTypeName = objectTypeName + "Args"
+					}
+					imports = append(imports, pulumiInputImport(pkg, module, objectTypeName))
+					for _, importDef := range collectObjectImports(object, objectType) {
+						imports = append(imports, importDef)
+					}
+				}
+			case *schema.ArrayType:
+				arrayType := inputType.(*schema.ArrayType)
+				innerArrayType := arrayType.ElementType
+				switch innerArrayType.(type) {
+				case *schema.ObjectType:
+					objectType := innerArrayType.(*schema.ObjectType)
+					switch inputProperty.Value.(type) {
+					case *model.TupleConsExpression:
+						objects := inputProperty.Value.(*model.TupleConsExpression)
+						for _, item := range objects.Expressions {
+							switch item.(type) {
+							case *model.ObjectConsExpression:
+								object := item.(*model.ObjectConsExpression)
+								fullyQualifiedTypeName := objectType.Token
+								nameParts := strings.Split(fullyQualifiedTypeName, ":")
+								objectTypeName := names.Title(nameParts[len(nameParts)-1])
+								if objectType.IsInputShape() {
+									objectTypeName = objectTypeName + "Args"
+								}
+								imports = append(imports, pulumiInputImport(pkg, module, objectTypeName))
+								for _, importDef := range collectObjectImports(object, objectType) {
+									imports = append(imports, importDef)
+								}
```

Comment:
It looks like this block repeats 5-6 times. Is there a reason why it couldn't be refactored out?

Reply:
Yeah I will try refactor into a separate method ✅